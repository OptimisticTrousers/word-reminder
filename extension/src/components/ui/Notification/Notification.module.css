.notification {
  border-radius: 6px;
  padding: 1rem;
}

.notification--success {
  --bkg-color: rgb(220, 252, 231);
  --base-text-color: rgb(20, 83, 45);
  --btn-text-color: rgb(22, 163, 74);
  --icon-color: rgb(21, 128, 61);
  background-color: var(--bkg-color);
}

.notification--error {
}

.notification__flex {
  display: flex;
  align-items: center;
}

.notification__item {
  flex-shrink: 0;
}

.notification__box {
  margin-left: 0.75rem;
}

.notification__message {
  color: var(--base-text-color);
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25rem;
  display: block;
}

.notification__controls {
  margin-left: auto;
  padding-left: 0.75rem;
}

.notification__control {
  margin-top: -0.375rem;
  margin-right: -0.375rem;
  margin-bottom: -0.375rem;
  margin-left: -0.375rem;
}

.notification__button {
  color: var(--btn-text-color);
  padding: 0.375rem;
  background-color: var(--bkg-color);
  border-radius: 6px;
  display: inline-flex;
}

.notification__text {
  position: absolute;
  width: 1px;
  padding: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

.notification__icon {
  width: 1.25rem;
  height: 1.25rem;
}

.notification__icon--circle {
  color: var(--icon-color);
}
