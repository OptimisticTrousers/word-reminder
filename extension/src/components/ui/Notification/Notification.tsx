import { CircleCheck, CircleX, X } from "lucide-react";
import { useContext } from "react";
import CSSModules from "react-css-modules";

import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import styles from "./Notification.module.css";

export const Notification = CSSModules(
  function ({ type, message }: { type: string; message: string }) {
    const { dismissNotification } = useContext(NotificationContext);

    function handleClick() {
      dismissNotification();
    }

    return (
      <div
        styleName={`notification notification--${type}`}
        role="dialog"
        aria-labelledby="notification-message"
      >
        <div styleName="notification__flex">
          <div styleName="notification__item">
            {type === NOTIFICATION_ACTIONS.SUCCESS ? (
              <CircleCheck
                data-testid="check-icon"
                styleName="notification__icon notification__icon--circle"
              />
            ) : (
              <CircleX
                data-testid="x-icon"
                styleName="notification__icon notification__icon--circle"
              />
            )}
          </div>
          <div styleName="notification__box">
            <span id="notification-message" styleName="notification__message">
              {message}
            </span>
          </div>
          <div styleName="notification__controls">
            <div styleName="notification__control">
              <button styleName="notification_button" onClick={handleClick}>
                <span styleName="notification__text">Dismiss</span>
                <X styleName="notification__icon" />
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
