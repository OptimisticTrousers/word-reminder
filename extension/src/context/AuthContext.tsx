import { createContext, FC, useReducer } from "react";

interface IUser {
  _id: string;
  username: string;
  theme: string;
  words: IUserWord[];
}

interface IWord {
  _id: string;
  word: string;
  origin: string;
  phonetic: string;
  audio: string;
  meanings: [
    {
      partOfSpeech: string;
      definitions: [
        {
          definition: string;
          example: string;
          synonyms: string[];
          antonyms: string[];
        }
      ];
    }
  ];
}

export interface IUserWord {
  _id: string;
  userId: IUser | string;
  word: IWord;
  learned: boolean;
}

interface AuthState {
  user: IUser | null;
}

interface Action {
  type: string;
  payload: IUser;
}

interface AuthContext {
  state: AuthState;
  dispatch: (action: Action) => void;
}

const authReducer = (state: AuthState, action: Action) => {
  switch (action.type) {
    // Login action will always contain a payload of user object
    case "LOGIN":
      return { user: action.payload };
    // Remove user object on logout action dispatch
    case "LOGOUT":
      return { user: null };
    default:
      return state;
  }
};

export const AuthContext = createContext<AuthContext>({} as AuthContext);

interface Props {
  children: JSX.Element[] | JSX.Element;
}

const AuthProvider: FC<Props> = ({ children }) => {
  const [state, dispatch] = useReducer(authReducer, {
    user: null,
  });

  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
