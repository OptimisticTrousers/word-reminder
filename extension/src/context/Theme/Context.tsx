import { createContext } from "react";

import { ThemeContext as Context } from "./Provider";

export enum Theme {
  Dark = "dark",
  Light = "light",
}

export const ThemeContext = createContext<Context>({
  theme: Theme.Dark,
  setTheme: function () {},
});
