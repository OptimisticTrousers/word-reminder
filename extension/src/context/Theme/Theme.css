.container {
  background-color: var(--bkg-color);
  color: var(--text-color);
}

.container--light {
  --bkg-color: #fff;
  --hover-bkg-color: rgb(12, 11, 9);
  --text-color: #222;
  --anchor-color: #0033cc;
  --input-bkg-color: #fff;
  --label-color: none;
  --shadow-color: none;
  --input-color: none;
  --primary-card-bkg-color: none;
  --secondary-card-bkg-color: none;
  --accordion-bkg-color: none;
  --control-bkg-color: none;
  --border-color: none;
  --icon-color: none;
  --navigation-bkg-color: none;
  --bottom-navigation-bkg-color: none;
  --active-navigation-bkg-color: none;
  --modal-bkg-color: rgba(245, 245, 245, 0.8);
  --break-color: #dedede;
  --secondary-icon-color: #606770;
  --hover-icon-color: rgb(59, 130, 246);
  --alert-text-color: rgb(185, 28, 28);
  --alert-bkg-color: rgb(254, 226, 226);
  --white-text-color: rgb(31, 41, 55);
  --white-bkg-color: rgb(243, 244, 246);
  --hover-white-bkg-color: rgb(229, 231, 235);
  --red-bkg-color: rgb(239, 68, 68);
  --hover-red-bkg-color: rgb(220, 38, 38);
}

.container--dark {
  --bkg-color: #242424;
  --text-color: #eee;
  --secondary-text-color: rgb(107, 114, 128);
  --tertiary-text-color: rgb(175, 170, 156);
  --anchor-color: #809fff;
  --secondary-anchor-color: #181819;
  --input-bkg-color: #2a2a2a;
  --label-color: #bac0ce;
  --shadow-color: #757575;
  --input-color: #a3a3a3;
  --card-bkg-color: #2a2a2a;
  --secondary-card-bkg-color: #161d3a;
  --accordion-bkg-color: #1f1e22;
  --top-navigation-bkg-color: #010001;
  --control-bkg-color: #2f343d;
  --border-color: #333;
  --icon-color: #efefef;
  --bottom-navigation-bkg-color: #f8f8f8;
  --active-navigation-bkg-color: #d3d3d3;
  --modal-bkg-color: rgba(42, 42, 42, 0.8);
  --break-color: #dedede;
  --secondary-icon-color: #606770;
  --hover-icon-color: rgb(59, 130, 246);
  --alert-text-color: rgb(185, 28, 28);
  --alert-bkg-color: rgb(254, 226, 226);
  --white-text-color: rgb(31, 41, 55);
  --white-bkg-color: rgb(243, 244, 246);
  --hover-white-bkg-color: rgb(229, 231, 235);
  --red-bkg-color: rgb(239, 68, 68);
  --hover-red-bkg-color: rgb(220, 38, 38);

  /* --color-control-bg: #2f343d;
  --color-word-bg: #161d3a;
  --color-base-text: white;
  --color-input-bg: #fff;
  --color-input-text: #a3a3a3;
  --color-button-bg: #7f8ff4;
  --color-button-text: #fff;
  --color-options-text: black;
  --color-accordion-bg: #1f1e22;
  --color-word-text: white;
  --color-accordion-text: white;
  --color-shadow: #757575;
  --color-label-text: #bac0ce; */
}

.container a {
  color: var(--anchor-color);
}

.container button,
.container input[type="button"] {
  padding: 0.6rem;
  border-radius: 8px;
  border: none;
  background-color: var(--primary-bkg-color);
  cursor: pointer;
  transition: background-color 0.3s ease;
}

.container button:hover,
.container input:hover[type="button"] {
  background-color: var(--primary-bkg-hover-color);
}

.container button:focus,
.container input:focus,
.container select:focus {
  outline: none;
  box-shadow: 0 0 5px var(--primary-color);
}

.container img {
  filter: brightness(0.8) contrast(1.2);
}

.container input {
  border: 1px solid var(--border-color);
}

.container input,
.container select,
.container fieldset {
  padding: 0.5rem;
  border-radius: 8px;
  background-color: var(--input-bkg-color);
}
