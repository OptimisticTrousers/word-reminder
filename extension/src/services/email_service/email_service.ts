import { Templates } from "common";

import { service } from "../service";

interface Body {
  email: string;
  subject: string;
  template: Templates;
}

export const emailService = (function (service) {
  const { post, VITE_API_DOMAIN } = service;

  function sendEmail(body: Body) {
    return post({
      url: `${VITE_API_DOMAIN}/emails`,
      options: {
        body: JSON.stringify(body),
        credentials: "include",
      },
    });
  }

  return { sendEmail };
})(service);
