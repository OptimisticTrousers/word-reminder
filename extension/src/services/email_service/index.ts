export { emailService } from "./email_service";
