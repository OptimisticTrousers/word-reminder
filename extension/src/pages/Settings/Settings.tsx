import { Subject, Template, User } from "common";
import { MouseEvent, useContext, useState } from "react";
import CSSModules from "react-css-modules";
import { useOutletContext } from "react-router-dom";

import styles from "./Settings.module.css";
import { useMutation } from "@tanstack/react-query";
import { emailService } from "../../services/email_service";
import { useNotificationError } from "../../hooks/useNotificationError";
import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../context/Notification";
import { DeleteUserModal } from "../../components/modals/DeleteUserModal";
import { ThemeContext } from "../../context/Theme";

export const Settings = CSSModules(
  function () {
    const { user }: { user: User } = useOutletContext();
    const userId = String(user.id);
    const { showNotificationError } = useNotificationError();
    const { showNotification } = useContext(NotificationContext);
    const { theme, toggleTheme } = useContext(ThemeContext);
    const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
    const { isPending, mutate } = useMutation({
      mutationFn: emailService.sendEmail,
      onSuccess: (_, variables) => {
        let field = "";
        switch (variables.body.template) {
          case Template.CHANGE_EMAIL:
            field = "email";
            break;
          case Template.CHANGE_PASSWORD:
            field = "password";
            break;
        }
        showNotification(
          NOTIFICATION_ACTIONS.SUCCESS,
          SETTINGS_NOTIFICATION_MSGS.sendEmail(field)
        );
      },
      onError: showNotificationError,
    });

    function toggleDeleteModal() {
      setIsDeleteModalOpen((prevValue) => !prevValue);
    }

    function handleEmailClick(event: MouseEvent<HTMLInputElement>) {
      const name = event.currentTarget.name;
      let template = Template.CHANGE_EMAIL;
      let subject = Subject.CHANGE_EMAIL;
      switch (name) {
        case "email":
          subject = Subject.CHANGE_EMAIL;
          template = Template.CHANGE_EMAIL;
          break;
        case "password":
          subject = Subject.CHANGE_PASSWORD;
          template = Template.CHANGE_PASSWORD;
          break;
      }
      mutate({
        userId,
        body: {
          email: user.email,
          subject,
          template,
        },
      });
    }

    const disabled = isPending;

    return (
      <section
        styleName={`settings settings--${theme}`}
        aria-labelledby="heading"
      >
        <h2 styleName="settings__heading" id="heading">
          Change Account Details
        </h2>
        <div styleName="settings__buttons">
          <div styleName="settings__control">
            <label styleName="settings__label" htmlFor="email">
              Change Email
            </label>
            <input
              id="email"
              type="button"
              name="email"
              styleName="settings__input"
              onClick={handleEmailClick}
              disabled={disabled}
            />
          </div>
          <div styleName="settings__control">
            <label styleName="settings__label" htmlFor="password">
              Change Password
            </label>
            <input
              id="password"
              type="button"
              name="password"
              styleName="settings__input"
              onClick={handleEmailClick}
              disabled={disabled}
            />
          </div>
          <div styleName="settings__control">
            <label styleName="settings__label" htmlFor="user">
              Delete User
            </label>
            <input
              id="user"
              type="button"
              name="user"
              styleName="settings__input"
              onClick={toggleDeleteModal}
              disabled={disabled}
            />
          </div>
          <div styleName="settings__control">
            <button
              styleName="settings__button"
              onClick={toggleTheme}
              role="switch"
            >
              Toggle Theme
            </button>
          </div>
        </div>
        {isDeleteModalOpen && (
          <DeleteUserModal toggleModal={toggleDeleteModal} />
        )}
      </section>
    );
  },
  styles,
  {
    allowMultiple: true,
    handleNotFoundStyleName: "log",
  }
);

const SETTINGS_NOTIFICATION_MSGS = {
  sendEmail: (field: string) => {
    return `A confirmation email was sent to your email to update your ${field}.`;
  },
};
