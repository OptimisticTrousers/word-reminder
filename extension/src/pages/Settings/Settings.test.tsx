import { render, screen } from "@testing-library/react";
import { Settings } from "./Settings";
import { createRoutesStub, Outlet } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { emailService } from "../../services/email_service";
import { Subject, Template } from "common";
import { NotificationProvider } from "../../context/Notification";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { Theme, ThemeContext } from "../../context/Theme/Context";

vi.mock("../../components/modals/DeleteUserModal", function () {
  return {
    DeleteUserModal: function () {
      return <div data-testid="delete-user-modal"></div>;
    },
  };
});

describe("Settings component", () => {
  function setup({ initialRoute }: { initialRoute: string }) {
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return (
            <QueryClientProvider client={queryClient}>
              <ThemeContext.Provider
                value={{ theme: Theme.Dark, toggleTheme: vi.fn() }}
              >
                <NotificationProvider>
                  <Outlet context={{ user: testUser }} />
                </NotificationProvider>
              </ThemeContext.Provider>
            </QueryClientProvider>
          );
        },
        children: [
          {
            path: "/settings",
            Component: Settings,
          },
        ],
      },
    ]);
    return {
      user: userEvent.setup(),
      ...render(<Stub initialEntries={[initialRoute]} />),
    };
  }

  const info = {
    message:
      "Message sent: <d786aa62-4e0a-070a-47ed-0b0666549519@ethereal.email",
  };

  const testUser = {
    email: "bob@gmail.com",
    password: "password",
    id: "1",
  };

  const status = 200;

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders non disabled inputs and heading", async () => {
    const { asFragment } = setup({ initialRoute: "/settings" });

    const heading = screen.getByRole("heading", {
      name: "Change Account Details",
    });
    const emailInput = screen.getByLabelText("Change Email", {
      selector: "input",
    });
    const passwordInput = screen.getByLabelText("Change Password", {
      selector: "input",
    });
    const deleteUserButton = screen.getByLabelText("Delete User", {
      selector: "input",
    });
    const toggleThemeButton = screen.getByRole("switch", {
      name: "Toggle Theme",
    });

    const section = screen.getByRole("region", {
      name: "Change Account Details",
    });
    expect(section).toBeInTheDocument();
    expect(section.getAttribute("class")).toContain(`settings--${Theme.Dark}`);
    expect(deleteUserButton).toBeInTheDocument();
    expect(toggleThemeButton).toBeInTheDocument();
    expect(heading).toBeInTheDocument();
    expect(emailInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("toggles theme to light mode", async () => {
    const { user } = setup({ initialRoute: "/settings" });

    const toggleThemeButton = screen.getByRole("switch", {
      name: "Toggle Theme",
    });
    await user.click(toggleThemeButton);

    const section = screen.getByRole("region", {
      name: "Change Account Details",
    });
    expect(section.getAttribute("class")).toContain(`settings--${Theme.Dark}`);
  });

  it("toggles theme to dark mode", async () => {
    const { user } = setup({ initialRoute: "/settings" });

    const toggleThemeButton = screen.getByRole("switch", {
      name: "Toggle Theme",
    });
    await user.click(toggleThemeButton);
    await user.click(toggleThemeButton);

    const section = screen.getByRole("region", {
      name: "Change Account Details",
    });
    expect(section.getAttribute("class")).toContain(`settings--${Theme.Dark}`);
  });

  it("opens the modal to delete the user's account when clicked", async () => {
    const { user } = setup({ initialRoute: "/settings" });

    const deleteUserButton = screen.getByRole("button", {
      name: "Delete User",
    });
    await user.click(deleteUserButton);

    const modal = screen.getByTestId("delete-user-modal");
    expect(modal).toBeInTheDocument();
  });

  it("sends change email verification email when clicking change email button", async () => {
    const mockSendEmail = vi
      .spyOn(emailService, "sendEmail")
      .mockImplementation(async () => {
        return { json: { info }, status };
      });
    const { user } = setup({ initialRoute: "/settings" });

    const changeEmailButton = screen.getByRole("button", {
      name: "Change Email",
    });
    await user.click(changeEmailButton);

    const notification = screen.getByRole("dialog", {
      name: "A confirmation email was sent to your email to update your email.",
    });
    expect(notification).toBeInTheDocument();
    expect(mockSendEmail).toHaveBeenCalledTimes(1);
    expect(mockSendEmail).toHaveBeenCalledWith({
      userId: testUser.id,
      body: {
        email: testUser.email,
        subject: Subject.CHANGE_EMAIL,
        template: Template.CHANGE_EMAIL,
      },
    });
  });

  it("sends change password verification email when clicking change password button", async () => {
    const mockSendEmail = vi
      .spyOn(emailService, "sendEmail")
      .mockImplementation(async () => {
        return { json: { info }, status };
      });
    const { user } = setup({ initialRoute: "/settings" });

    const changePasswordButton = screen.getByRole("button", {
      name: "Change Password",
    });
    await user.click(changePasswordButton);

    const notification = screen.getByRole("dialog", {
      name: "A confirmation email was sent to your email to update your password.",
    });
    expect(notification).toBeInTheDocument();
    expect(mockSendEmail).toHaveBeenCalledTimes(1);
    expect(mockSendEmail).toHaveBeenCalledWith({
      userId: testUser.id,
      body: {
        email: testUser.email,
        subject: Subject.CHANGE_PASSWORD,
        template: Template.CHANGE_PASSWORD,
      },
    });
  });

  it("calls the functions to show a notification error when changing email", async () => {
    const message = "Bad Request.";
    const status = 400;
    const mockSendEmail = vi
      .spyOn(emailService, "sendEmail")
      .mockImplementation(async () => {
        return Promise.reject({ json: { message }, status });
      });
    const { user } = setup({ initialRoute: "/settings" });

    const changeEmailButton = screen.getByRole("button", {
      name: "Change Email",
    });
    await user.click(changeEmailButton);

    const notification = screen.getByRole("dialog", { name: message });
    expect(notification).toBeInTheDocument();
    expect(mockSendEmail).toHaveBeenCalledTimes(1);
    expect(mockSendEmail).toHaveBeenCalledWith({
      userId: testUser.id,
      body: {
        email: testUser.email,
        subject: Subject.CHANGE_EMAIL,
        template: Template.CHANGE_EMAIL,
      },
    });
  });

  it("calls the functions to show a notification error when changing password", async () => {
    const message = "Bad Request.";
    const status = 400;
    const mockSendEmail = vi
      .spyOn(emailService, "sendEmail")
      .mockImplementation(async () => {
        return Promise.reject({ json: { message }, status });
      });
    const { user } = setup({ initialRoute: "/settings" });

    const changePasswordButton = screen.getByRole("button", {
      name: "Change Password",
    });
    await user.click(changePasswordButton);

    const notification = screen.getByRole("dialog", { name: message });
    expect(notification).toBeInTheDocument();
    expect(mockSendEmail).toHaveBeenCalledTimes(1);
    expect(mockSendEmail).toHaveBeenCalledWith({
      userId: testUser.id,
      body: {
        email: testUser.email,
        subject: Subject.CHANGE_PASSWORD,
        template: Template.CHANGE_PASSWORD,
      },
    });
  });
});
