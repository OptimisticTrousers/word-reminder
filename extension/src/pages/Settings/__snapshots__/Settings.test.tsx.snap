// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Settings component > renders non disabled inputs and heading 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="heading"
    class="_settings_9f5ad4 _settings--dark_9f5ad4"
  >
    <h2
      class="_settings__heading_9f5ad4"
      id="heading"
    >
      Change Account Details
    </h2>
    <div
      class="_settings__buttons_9f5ad4"
    >
      <div
        class="_settings__control_9f5ad4"
      >
        <label
          class="_settings__label_9f5ad4"
          for="email"
        >
          Change Email
        </label>
        <input
          class="_settings__input_9f5ad4"
          id="email"
          name="email"
          type="button"
        />
      </div>
      <div
        class="_settings__control_9f5ad4"
      >
        <label
          class="_settings__label_9f5ad4"
          for="password"
        >
          Change Password
        </label>
        <input
          class="_settings__input_9f5ad4"
          id="password"
          name="password"
          type="button"
        />
      </div>
      <div
        class="_settings__control_9f5ad4"
      >
        <label
          class="_settings__label_9f5ad4"
          for="user"
        >
          Delete User
        </label>
        <input
          class="_settings__input_9f5ad4"
          id="user"
          name="user"
          type="button"
        />
      </div>
      <div
        class="_settings__control_9f5ad4"
      >
        <button
          class="_settings__button_9f5ad4"
          role="switch"
        >
          Toggle Theme
        </button>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
